File: packages/firebase_auth/firebase_auth/ios/firebase_auth/Sources/firebase_auth/FLTIdTokenChannelStreamHandler.m
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
@import FirebaseAuth;
#import "include/Private/FLTIdTokenChannelStreamHandler.h"
#import <FirebaseAuth/FirebaseAuth.h>
#import "include/Private/PigeonParser.h"
#import "include/Public/FLTFirebaseAuthPlugin.h"

@implementation FLTIdTokenChannelStreamHandler {
  FIRAuth *_auth;
  FIRIDTokenDidChangeListenerHandle _listener;
}

- (instancetype)initWithAuth:(FIRAuth *)auth {
  self = [super init];
  if (self) {
    _auth = auth;
  }
  return self;
}

- (FlutterError *)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)events {
  bool __block initialAuthState = YES;

  _listener = [_auth addIDTokenDidChangeListener:^(FIRAuth *_Nonnull auth,
                                                   FIRUser *_Nullable user) {
    if (initialAuthState) {
      initialAuthState = NO;
      return;
    }

    if (user) {
      events(@{
        @"user" : [PigeonParser getManualList:[PigeonParser getPigeonDetails:[auth currentUser]]]
      });
    } else {
      events(@{@"user" : [NSNull null]});
    }
  }];

  return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments {
  if (_listener) {
    [_auth removeIDTokenDidChangeListener:_listener];
  }
  _listener = nil;

  return nil;
}

@end

File: packages/firebase_auth/firebase_auth/ios/firebase_auth/Sources/firebase_auth/FLTIdTokenChannelStreamHandler.swift
// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import FirebaseAuth
import Foundation

#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#endif

final class FLTIdTokenChannelStreamHandler: NSObject, FlutterStreamHandler {
  private let auth: Auth
  private var handle: IDTokenDidChangeListenerHandle?

  init(auth: Auth) {
    self.auth = auth
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    var initialAuthState = true
    handle = auth.addIDTokenDidChangeListener { auth, user in
      if initialAuthState {
        initialAuthState = false
        return
      }

      if user != nil, let currentUser = auth.currentUser {
        events(["user": PigeonParser.getManualList(PigeonParser.getPigeonDetails(currentUser))])
      } else {
        events(["user": NSNull()])
      }
    }
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    if let handle {
      auth.removeIDTokenDidChangeListener(handle)
    }
    handle = nil
    return nil
  }
}
